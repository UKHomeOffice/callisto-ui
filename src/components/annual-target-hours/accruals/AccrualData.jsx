import PropTypes from 'prop-types';
import { useTranslation } from 'react-i18next';
import '../../../i18n';
import { formatToHoursAndMinutes } from '../../../utils/time-entry-utils/timeEntryUtils';

const AccrualData = ({ targetData, accrualsData, titleTranslationKey }) => {
  const { t } = useTranslation('common');
  const total = targetData?.targetTotal || 0;
  const worked = accrualsData?.cumulativeTotal || 0;
  const remainingMins = Math.floor(total - worked);
  const remainingHours = Math.floor(remainingMins / 60);
  const target = total - (accrualsData?.cumulativeTarget || 0);
  const idSubstr = titleTranslationKey?.split('.')[0];
  let title;

  if (!targetData) {
    title = t('accrualsData.noAgreement');
  } else {
    title = t(titleTranslationKey, {
      count: remainingHours,
    });
  }

  return (
    <div className="accruals-container">
      <form className="grey-border">
        <h1 className="govuk-heading-m newline">
          <span
            id={`${idSubstr}-title-span`}
            dangerouslySetInnerHTML={{
              __html: title,
            }}
          />
        </h1>
        <table className="govuk-table">
          <tbody className="govuk-table__body">
            <tr className="govuk-table__row">
              <th
                id={`${idSubstr}-total-lbl`}
                scope="row"
                className="govuk-table__header"
              >
                {t('accrualsData.total')}
              </th>
              <td
                id={`${idSubstr}-total-value`}
                className="govuk-table__cell govuk-table__cell--numeric"
              >
                {targetData ? formatToHoursAndMinutes(total) : '-'}
              </td>
            </tr>
            <tr className="govuk-table__row">
              <th
                id={`${idSubstr}-worked-lbl`}
                scope="row"
                className="govuk-table__header"
              >
                {t('accrualsData.worked')}
              </th>
              <td
                id={`${idSubstr}-worked-value`}
                className="govuk-table__cell govuk-table__cell--numeric"
              >
                {targetData ? formatToHoursAndMinutes(worked) : '-'}
              </td>
            </tr>
            <tr className="govuk-table__row">
              <th
                id={`${idSubstr}-remaining-lbl`}
                scope="row"
                className="govuk-table__header"
              >
                {t('accrualsData.remaining')}
              </th>
              <td
                id={`${idSubstr}-remaining-value`}
                className="govuk-table__cell govuk-table__cell--numeric"
              >
                {targetData ? formatToHoursAndMinutes(remainingMins) : '-'}
              </td>
            </tr>
            <tr className="govuk-table__row">
              <th
                id={`${idSubstr}-target-lbl`}
                scope="row"
                className="govuk-table__header"
              >
                {t('accrualsData.target')}
              </th>
              <td
                id={`${idSubstr}-target-value`}
                className="govuk-table__cell govuk-table__cell--numeric"
              >
                {targetData ? formatToHoursAndMinutes(target) : '-'}
              </td>
            </tr>
          </tbody>
        </table>
      </form>
    </div>
  );
};

export default AccrualData;

AccrualData.propTypes = {
  targetData: PropTypes.shape({
    id: PropTypes.string,
    tenantId: PropTypes.string,
    agreementId: PropTypes.string,
    accrualTypeId: PropTypes.string,
    targetTotal: PropTypes.number,
  }),
  accrualsData: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string,
        tenantId: PropTypes.string,
        personId: PropTypes.string,
        agreementId: PropTypes.string,
        accrualDate: PropTypes.string,
        accrualTypeId: PropTypes.string,
        cumulativeTotal: PropTypes.number,
        cumulativeTarget: PropTypes.number,
        contributions: PropTypes.shape({
          timeEntries: PropTypes.object,
          total: PropTypes.number,
        }),
      })
    ),
  ]),
  titleTranslationKey: PropTypes.string,
};
