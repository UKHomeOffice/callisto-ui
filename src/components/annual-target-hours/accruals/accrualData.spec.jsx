import { render, screen } from '@testing-library/react';
import AccrualData from './AccrualData';
import pretty from 'pretty';

describe('AccrualData', () => {
  it('should render a table with correct headers', () => {
    const { getByRole, baseElement } = render(
      <AccrualData targetData={null} accrualsData={null} />
    );

    const table = getByRole('table');
    expect(table).toBeTruthy();

    const headers = ['Total', 'Worked', 'Remaining', 'Target'];

    headers.forEach((header) => {
      expect(screen.getByText(header)).toBeTruthy();
    });
    expect(pretty(baseElement.innerHTML)).toMatchSnapshot();
  });

  it('should display correct title for annual Target Hours', () => {
    const targetData = { targetTotal: 131520 };
    const accrualsData = { cumulativeTotal: 600 };
    const titleTranslationKey = 'annualTargetHours.remainingHoursTitle';

    const { baseElement } = render(
      <AccrualData
        targetData={targetData}
        accrualsData={accrualsData}
        titleTranslationKey={titleTranslationKey}
      />
    );

    expect(screen.getByText('Annual target hours remaining')).toBeTruthy();
    expect(screen.getByText('2182')).toBeTruthy();
    expect(pretty(baseElement.innerHTML)).toMatchSnapshot();
  });

  it('should display correct title for night Hours', () => {
    const targetData = { targetTotal: 131520 };
    const accrualsData = { cumulativeTotal: 600 };
    const titleTranslationKey = 'nightHours.remainingHoursTitle';

    const { baseElement } = render(
      <AccrualData
        targetData={targetData}
        accrualsData={accrualsData}
        titleTranslationKey={titleTranslationKey}
      />
    );

    expect(screen.getByText('Night hours remaining')).toBeTruthy();
    expect(screen.getByText('2182')).toBeTruthy();
    expect(pretty(baseElement.innerHTML)).toMatchSnapshot();
  });
  it('should display correct text for remaining hours without targetData', () => {
    const targetData = null;
    const accrualsData = { cumulativeTotal: 500 };

    const { baseElement } = render(
      <AccrualData targetData={targetData} accrualsData={accrualsData} />
    );

    const expectedText = 'No agreement has been found';

    expect(screen.getByText(expectedText)).toBeTruthy();
    expect(pretty(baseElement.innerHTML)).toMatchSnapshot();
  });

  it('should display "-" for headers without targetData', () => {
    const targetData = null;
    const accrualsData = { cumulativeTotal: 100 };

    const { baseElement } = render(
      <AccrualData targetData={targetData} accrualsData={accrualsData} />
    );

    const expectedText = screen.getAllByText('-');

    expect(expectedText.length).toBe(4);
    expect(pretty(baseElement.innerHTML)).toMatchSnapshot();
  });
});
