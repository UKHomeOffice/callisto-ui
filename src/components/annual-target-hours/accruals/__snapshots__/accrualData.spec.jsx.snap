// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccrualData should display "-" for headers without targetData 1`] = `
"<div>
  <div class=\\"accruals-container\\">
    <form class=\\"grey-border\\">
      <h1 class=\\"govuk-heading-m newline\\">No agreement has been found</h1>
      <table class=\\"govuk-table\\">
        <tbody class=\\"govuk-table__body\\">
          <tr class=\\"govuk-table__row\\">
            <th data-testid=\\"null-total-lbl\\" scope=\\"row\\" class=\\"govuk-table__header\\">Total</th>
            <td data-testid=\\"null-total-value\\" class=\\"govuk-table__cell govuk-table__cell--numeric\\">-</td>
          </tr>
          <tr class=\\"govuk-table__row\\">
            <th data-testid=\\"null-worked-lbl\\" scope=\\"row\\" class=\\"govuk-table__header\\">Worked</th>
            <td data-testid=\\"null-worked-value\\" class=\\"govuk-table__cell govuk-table__cell--numeric\\">-</td>
          </tr>
          <tr class=\\"govuk-table__row\\">
            <th data-testid=\\"null-remaining-lbl\\" scope=\\"row\\" class=\\"govuk-table__header\\">Remaining</th>
            <td data-testid=\\"null-remaining-value\\" class=\\"govuk-table__cell govuk-table__cell--numeric\\">-</td>
          </tr>
          <tr class=\\"govuk-table__row\\">
            <th data-testid=\\"null-target-lbl\\" scope=\\"row\\" class=\\"govuk-table__header\\">Target</th>
            <td data-testid=\\"null-target-value\\" class=\\"govuk-table__cell govuk-table__cell--numeric\\">-</td>
          </tr>
        </tbody>
      </table>
    </form>
  </div>
</div>"
`;

exports[`AccrualData should display correct text for remaining hours without targetData 1`] = `
"<div>
  <div class=\\"accruals-container\\">
    <form class=\\"grey-border\\">
      <h1 class=\\"govuk-heading-m newline\\">No agreement has been found</h1>
      <table class=\\"govuk-table\\">
        <tbody class=\\"govuk-table__body\\">
          <tr class=\\"govuk-table__row\\">
            <th data-testid=\\"null-total-lbl\\" scope=\\"row\\" class=\\"govuk-table__header\\">Total</th>
            <td data-testid=\\"null-total-value\\" class=\\"govuk-table__cell govuk-table__cell--numeric\\">-</td>
          </tr>
          <tr class=\\"govuk-table__row\\">
            <th data-testid=\\"null-worked-lbl\\" scope=\\"row\\" class=\\"govuk-table__header\\">Worked</th>
            <td data-testid=\\"null-worked-value\\" class=\\"govuk-table__cell govuk-table__cell--numeric\\">-</td>
          </tr>
          <tr class=\\"govuk-table__row\\">
            <th data-testid=\\"null-remaining-lbl\\" scope=\\"row\\" class=\\"govuk-table__header\\">Remaining</th>
            <td data-testid=\\"null-remaining-value\\" class=\\"govuk-table__cell govuk-table__cell--numeric\\">-</td>
          </tr>
          <tr class=\\"govuk-table__row\\">
            <th data-testid=\\"null-target-lbl\\" scope=\\"row\\" class=\\"govuk-table__header\\">Target</th>
            <td data-testid=\\"null-target-value\\" class=\\"govuk-table__cell govuk-table__cell--numeric\\">-</td>
          </tr>
        </tbody>
      </table>
    </form>
  </div>
</div>"
`;

exports[`AccrualData should display correct title for annual Target Hours 1`] = `
"<div>
  <div class=\\"accruals-container\\">
    <form class=\\"grey-border\\">
      <h1 class=\\"govuk-heading-m newline\\"><span data-testid=\\"annualTargetHours-title-span\\">Annual target hours
 <strong>2182</strong> remaining</span></h1>
      <table class=\\"govuk-table\\">
        <tbody class=\\"govuk-table__body\\">
          <tr class=\\"govuk-table__row\\">
            <th data-testid=\\"annualTargetHours-total-lbl\\" scope=\\"row\\" class=\\"govuk-table__header\\">Total</th>
            <td data-testid=\\"annualTargetHours-total-value\\" class=\\"govuk-table__cell govuk-table__cell--numeric\\">2192:00</td>
          </tr>
          <tr class=\\"govuk-table__row\\">
            <th data-testid=\\"annualTargetHours-worked-lbl\\" scope=\\"row\\" class=\\"govuk-table__header\\">Worked</th>
            <td data-testid=\\"annualTargetHours-worked-value\\" class=\\"govuk-table__cell govuk-table__cell--numeric\\">10:00</td>
          </tr>
          <tr class=\\"govuk-table__row\\">
            <th data-testid=\\"annualTargetHours-remaining-lbl\\" scope=\\"row\\" class=\\"govuk-table__header\\">Remaining</th>
            <td data-testid=\\"annualTargetHours-remaining-value\\" class=\\"govuk-table__cell govuk-table__cell--numeric\\">2182:00</td>
          </tr>
          <tr class=\\"govuk-table__row\\">
            <th data-testid=\\"annualTargetHours-target-lbl\\" scope=\\"row\\" class=\\"govuk-table__header\\">Target</th>
            <td data-testid=\\"annualTargetHours-target-value\\" class=\\"govuk-table__cell govuk-table__cell--numeric\\">2192:00</td>
          </tr>
        </tbody>
      </table>
    </form>
  </div>
</div>"
`;

exports[`AccrualData should display correct title for night Hours 1`] = `
"<div>
  <div class=\\"accruals-container\\">
    <form class=\\"grey-border\\">
      <h1 class=\\"govuk-heading-m newline\\"><span data-testid=\\"nightHours-title-span\\">Night hours
 <strong>2182</strong> remaining</span></h1>
      <table class=\\"govuk-table\\">
        <tbody class=\\"govuk-table__body\\">
          <tr class=\\"govuk-table__row\\">
            <th data-testid=\\"nightHours-total-lbl\\" scope=\\"row\\" class=\\"govuk-table__header\\">Total</th>
            <td data-testid=\\"nightHours-total-value\\" class=\\"govuk-table__cell govuk-table__cell--numeric\\">2192:00</td>
          </tr>
          <tr class=\\"govuk-table__row\\">
            <th data-testid=\\"nightHours-worked-lbl\\" scope=\\"row\\" class=\\"govuk-table__header\\">Worked</th>
            <td data-testid=\\"nightHours-worked-value\\" class=\\"govuk-table__cell govuk-table__cell--numeric\\">10:00</td>
          </tr>
          <tr class=\\"govuk-table__row\\">
            <th data-testid=\\"nightHours-remaining-lbl\\" scope=\\"row\\" class=\\"govuk-table__header\\">Remaining</th>
            <td data-testid=\\"nightHours-remaining-value\\" class=\\"govuk-table__cell govuk-table__cell--numeric\\">2182:00</td>
          </tr>
          <tr class=\\"govuk-table__row\\">
            <th data-testid=\\"nightHours-target-lbl\\" scope=\\"row\\" class=\\"govuk-table__header\\">Target</th>
            <td data-testid=\\"nightHours-target-value\\" class=\\"govuk-table__cell govuk-table__cell--numeric\\">2192:00</td>
          </tr>
        </tbody>
      </table>
    </form>
  </div>
</div>"
`;

exports[`AccrualData should render a table with correct headers 1`] = `
"<div>
  <div class=\\"accruals-container\\">
    <form class=\\"grey-border\\">
      <h1 class=\\"govuk-heading-m newline\\">No agreement has been found</h1>
      <table class=\\"govuk-table\\">
        <tbody class=\\"govuk-table__body\\">
          <tr class=\\"govuk-table__row\\">
            <th data-testid=\\"null-total-lbl\\" scope=\\"row\\" class=\\"govuk-table__header\\">Total</th>
            <td data-testid=\\"null-total-value\\" class=\\"govuk-table__cell govuk-table__cell--numeric\\">-</td>
          </tr>
          <tr class=\\"govuk-table__row\\">
            <th data-testid=\\"null-worked-lbl\\" scope=\\"row\\" class=\\"govuk-table__header\\">Worked</th>
            <td data-testid=\\"null-worked-value\\" class=\\"govuk-table__cell govuk-table__cell--numeric\\">-</td>
          </tr>
          <tr class=\\"govuk-table__row\\">
            <th data-testid=\\"null-remaining-lbl\\" scope=\\"row\\" class=\\"govuk-table__header\\">Remaining</th>
            <td data-testid=\\"null-remaining-value\\" class=\\"govuk-table__cell govuk-table__cell--numeric\\">-</td>
          </tr>
          <tr class=\\"govuk-table__row\\">
            <th data-testid=\\"null-target-lbl\\" scope=\\"row\\" class=\\"govuk-table__header\\">Target</th>
            <td data-testid=\\"null-target-value\\" class=\\"govuk-table__cell govuk-table__cell--numeric\\">-</td>
          </tr>
        </tbody>
      </table>
    </form>
  </div>
</div>"
`;
